Add autoexport to automatically configure trace exporters

MikeGoldsmith opened the pull request:
This PR adds an autoexport package similar to autoprop that can detect and configure trace exporters, or use provided ones if none are set in environment variables.

Note: common exporters (eg otlp, jaeger, etc) cannot be pre-populated into the registry during init because they would not take into account environment variables configured before autoexport is called. This is valuable, as it both supports testing and when combined with the proposed generic config launcher (see below).

- #2592 
- https://github.com/honeycombio/opentelemetry-go-contrib/pull/400

Known limitations:
- Only one exporter per type can be configured, eg a single "otlp" exporter can be configured
- Does not consider metrics exporters yet because they are slightly more complicated because metrics exporters are wrapped by a controller (push / pull)

MrAlias commented on exporters/autoexport/exporter.go:
```diff
@@ -0,0 +1,95 @@
+	"go.opentelemetry.io/otel/sdk/trace"
+)
+
+const otelTracesExportersEnvKey = "OTEL_EXPORTERS"
```
```suggestion
const otelTracesExportersEnvKey = "OTEL_TRACES_EXPORTER"
```

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#exporter-selection

MrAlias commented on exporters/autoexport/exporter.go:
```diff
@@ -0,0 +1,95 @@
+const otelTracesExportersEnvKey = "OTEL_EXPORTERS"
+
+// errUnknownExpoter is returned when an unknown exporter name is used in
+// the OTEL_EXPORTERS environment variable.
```
```suggestion
// the OTEL_*_EXPORTER environment variables.
```

MrAlias commented on exporters/autoexport/exporter.go:
```diff
@@ -0,0 +1,95 @@
+// the OTEL_EXPORTERS environment variable.
+var errUnknownExpoter = errors.New("unknown exporter")
+
+func NewTraceExporters(exporters ...trace.SpanExporter) []trace.SpanExporter {
```
I think this needs to be defined for a single exporter.

```suggestion
func NewTraceExporter(exporter trace.SpanExporter) trace.SpanExporter {
```

That way it can be used similar to:

```go
tp := trace.NewTracerProvider(trace.WithBatcher(autoexport.NewTraceExporter(/* some defualt */)))
// ...
```

MikeGoldsmith replied:
Yeah, I agree. If we limit autoexport to only ever configure one exporter, it's a lot easier to reason about and removes the known limitation in the description (configuring multiple exporters with the same type).

MikeGoldsmith replied:
I've updated the func to only return a single SpanExporter, but the interface doesn't feel as nice. It was nice that the you could call autoexport.NewTraceExporter without a default exporter.

MrAlias commented on exporters/autoexport/go.mod:
```diff
@@ -0,0 +1,31 @@
+module go.opentelemetry.io/contrib/exporters/autoexport
+
+go 1.17
```
```suggestion
go 1.18
```

MrAlias commented on exporters/autoexport/registry.go:
```diff
@@ -0,0 +1,90 @@
+// RegisterSpanExporter sets the SpanExporter e to be used when the
+// OTEL_EXPORTERS environment variable contains the exporter name. This
+// will panic if name has already been registered.
+func RegisterSpanExporter(name string, e trace.SpanExporter) {
```
Similar to the [autoprop pacakge](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/c4876c33a8e2e4b09501013f3543153d8aa10178/propagators/autoprop/registry.go#L132), we'll probably want to add a `SpanExporter(name string) trace.SpanExporter` function for things like the otel launcher to use.

MikeGoldsmith replied:
Done 👍🏻

MikeGoldsmith commented:
Hey @MrAlias - could you give this another pass for feedback please? Thanks

MadVikingGod commented on exporters/autoexport/registry.go:
```diff
@@ -0,0 +1,138 @@
+// RegisterSpanExporter sets the SpanExporter e to be used when the
+// OTEL_TRACES_EXPORTERS environment variable contains the exporter name. This
+// will panic if name has already been registered.
+func RegisterSpanExporter(name string, e trace.SpanExporter) {
```
I know Start is only an OTLP exporter concept, but do you see any issues that might arise from packages registering a "started" client that isn't used?

eg. If I register a jaeger, an otlp, and a zipkin with the intent of the operator choosing via environment variables, will having misconfigured other exporters sitting in the registry cause problems?

MikeGoldsmith replied:
I don't think so, started exporters aren't actually wired up into a pipeline as part of this package and an exporter should only act on data to export. Started exporters may hold resources that would be idle and unused.

MikeGoldsmith replied:
Sorry I never got back to this @dashpole.

I think a factory would be useful to defer the span exporter creation, though it would mean we'd need to maintain a list of factories alongside the tranceExporter instances too. 

Maybe should can look at doing that as a follow-up optimsation, I don't think it's needed to start with.

pdelewski replied:
> Exporters are not expected to be concurrent safe and if we return the same exporter to all requests it may be they are registered with different span processors. I think for this reason the factor makes more sense here as we want to return a unique instance.

Maybe we can solve problem of uniqueness with adding additional properties to `envRegistry map` key? I would just avoid another level of indirection (factory in this case) until it's really necessary

TylerHelmuth commented on exporters/autoexport/exporter.go:
```diff
@@ -0,0 +1,55 @@
+	otelTracesExportersEnvKey = "OTEL_TRACES_EXPORTER"
+)
+
+// NewTraceExporter returns a conbfigured SpanExporter defined using the environment
```
```suggestion
// NewTraceExporter returns a configured SpanExporter defined using the environment
```

TylerHelmuth commented on exporters/autoexport/registry.go:
```diff
@@ -0,0 +1,138 @@
+	return p, ok
+}
+
+var errDupReg = errors.New("duplicate registration")
```
Should this be moved up into the vars section with the rest of the errors?

TylerHelmuth commented on exporters/autoexport/registry.go:
```diff
@@ -0,0 +1,138 @@
+//
+// An error is returned for any unknown exporters.
+func SpanExporter(name string) (trace.SpanExporter, error) {
+	if name == "otlp" {
```
Instead of having this default case, is it possible to initialize the registry with the grpc and http/protobuff otlp exporters during initialization?  Or does this logic end up being more efficient than storing otlp exporters in the registry?

MikeGoldsmith replied:
This does delay the exporter creation and therefor it's resources until it's selected as the exporter to be used which is nice. eg you won't end up with unused exporters in the registry.

Another benefit of lazily creating them like this is that we can test expected behaviour by setting env vars in unit tests before calling retrieving an exporter. We wouldn't get the opportunity to test the different exporter options if they were created on package initialisation, .

MikeGoldsmith replied:
I've updated the logic to register the OTLP exporter during `init()`. We only need one registration for an OTLP exporter, but under both `""` and `"otlp"` keys.

JamieDanielson commented on exporters/autoexport/exporter.go:
```diff
@@ -0,0 +1,55 @@
+// NewTraceExporter returns a configured SpanExporter defined using the environment
+// variable OTEL_TRACES_EXPORTER or the passed in exporter. The exporter defined
+// in OTEL_TRACES_EXPORTER is preferred over the exporter passed in.
+func NewTraceExporter(exporter trace.SpanExporter) trace.SpanExporter {
```
I think we want to have a default set here, an OTLP exporter with gRPC as default protocol. If nil provided, default to otlp/grpc, otherwise use what is passed in here.

Then in the check,

```go
if exporter == nil {
    return grpc // (default, however we define that) 
}
return exporter
```

MadVikingGod reviewed (approved):
LGTM.

Finish addressing the concerns others raised.

jmacd commented:
This seems fundamentally different than the `autoprop` package in the sense that it supports registration of span Exporters (a part of a provider), whereas the thing configured in autoprop is a whole provider of propagation functionality. When I consider how to auto-configure a Span exporter, that includes the notion of auto-configuring the processor at the same time. Some span exporters will want to use the batch processor, and some will not--although these are independent interfaces I do not believe they should be auto-configured separately.

So, the thing I expect a registry for auto-configured-span-exporters to return is a `TracerProviderOption` which encapsulates the processor and exporter (i.e., a whole pipeline). Likewise for auto-configuring metrics exporteres, I expect a registry for auto-configured-metric-exporters to configure `MetricReader` instances appropriately for some exporter and return a `MeterProviderOption`.

jmacd commented:
I'm just feeling wary of building automatic-configurators for individual building blocks of the SDK on the assumption that they are always perfectly orthogonal from one another. I would prefer instead of configuring Samplers, Processors, and Exporters independently (for spans) for us to configure feature-level units of code which sometimes means configuring more than one SDK interface with a coordinated implementation. [Here is an example of a combined java-contrib Sampler and Processor implementation.](https://github.com/open-telemetry/opentelemetry-java-contrib/pull/352)

MrAlias commented on exporters/autoexport/registry.go:
```diff
@@ -0,0 +1,139 @@
+// already registered SpanExporter or a default (otlp).
+//
+// An error is returned for any unknown exporters.
+func SpanExporter(name string) (trace.SpanExporter, error) {
```
If they support for configuration, the value could come from there. For an example: https://github.com/open-telemetry/opentelemetry-go-contrib/issues/2590

dashpole replied:
Thanks!

MikeGoldsmith commented:
Thanks for the feedback all, I'll be following up on queries and addressing feedback over the next couple of days.

MikeGoldsmith commented on exporters/autoexport/exporter_test.go:
```diff
@@ -0,0 +1,42 @@
+func TestExportersConfiguredInEnvVarAreReturned(t *testing.T) {
+	t.Setenv("OTEL_TRACES_EXPORTERS", "otlp")
+	exporter := NewTraceExporter(nil)
+	assert.NotNil(t, 1, exporter)
```
Fixed

MikeGoldsmith commented on exporters/autoexport/registry.go:
```diff
@@ -0,0 +1,139 @@
+// OTEL_TRACES_EXPORTERS environment variable contains the exporter name. This
+// will panic if name has already been registered.
+func RegisterSpanExporter(name string, e trace.SpanExporter) {
+	if err := envRegistry.store(name, e); err != nil {
```
Updated registry to be simpler and register the OTLP exporter during module init.

MikeGoldsmith commented on exporters/autoexport/exporter.go:
```diff
@@ -0,0 +1,55 @@
+// NewTraceExporter returns a configured SpanExporter defined using the environment
+// variable OTEL_TRACES_EXPORTER or the passed in exporter. The exporter defined
+// in OTEL_TRACES_EXPORTER is preferred over the exporter passed in.
+func NewTraceExporter(exporter trace.SpanExporter) trace.SpanExporter {
```
I've updated to use the option pattern. I agree it feels nicer as there are 3 levels of config (env, fallback, default).

MikeGoldsmith commented:
I've gone through and addressed all feedback.

Main changes are:
- Used Options pattern for providing fallback exporter
- Updated registry to populate with OTLP exporter during `init()` and simplified retrieval logic
- Added autoexport to dependabot and versions
  - @MrAlias I added to experimental-instrumenations as autoprop was there, let me know if you'd prefer somewhere else
  
Thanks all 😄

MikeGoldsmith commented:
Hmm - I've had another thought regarding using this package with the proposed go contrib launcher, and am wondering if moving the otlp exporter creation to init is a good idea. It means a vendor who wants to modify settings, eg override the exporter's endpoint, cannot do it before its created.

I think we need to delay creating the default (otlp) exporter until it's requested by calling `SpanExporter()`.

MikeGoldsmith commented:
> Hmm - I've had another thought regarding using this package with the proposed go contrib launcher, and am wondering if moving the otlp exporter creation to init is a good idea. It means a vendor who wants to modify settings, eg override the exporter's endpoint, cannot do it before its created.
> 
> I think we need to delay creating the default (otlp) exporter until it's requested by calling `SpanExporter()`.

Thinking more on this, in the scenario someone wants to alter the default OTLP exporter programatically (eg headers, endpoint, etc) - would we expect them to register a new exporter? Many users of the autoexport package will likely want to alter some settings.

For the launcher when using the OTLP exporter, if we want to edit the endpoint, we'd then have to create a new exporter anyway so makes no sense for using the autoexport package at all.

MikeGoldsmith commented:
> Hmm - I've had another thought regarding using this package with the proposed go contrib launcher, and am wondering if moving the otlp exporter creation to init is a good idea. It means a vendor who wants to modify settings, eg override the exporter's endpoint, cannot do it before its created.
> 
> I think we need to delay creating the default (otlp) exporter until it's requested by calling `SpanExporter()`.

Thinking more on this, in the scenario someone wants to alter the default OTLP exporter programatically (eg headers, endpoint, etc) - would we expect them to register a new exporter? Many users of the autoexport package will likely want to alter some settings.

For the launcher when using the OTLP exporter, if we want to edit the endpoint, we'd then have to create a new exporter anyway so makes no sense for using the autoexport package at all.

MikeGoldsmith commented:
I've talked myself round in a circle. I think what we have is good. If app, or an intermediate layer such as the launcher, wants to provide an alternative exporter configuration other than using env vars, then you should build the exporter and register it 👍🏻

MikeGoldsmith commented:
@open-telemetry/go-approvers @open-telemetry/go-maintainers  looking for another round of feedback on this PR please 👍🏻

alexrudd commented:
Hey, are there any updates on this? Would be a really great feature :)

pdelewski commented on exporters/autoexport/exporter.go:
```diff
@@ -0,0 +1,84 @@
+// parseEnv returns a configured SpanExporter defined by the OTEL_TRACES_EXPORTER
+// environment variable.
+// nil is returned if no exporter is defined for the environment variable.
+func parseEnv() (trace.SpanExporter, error) {
```
function name is misleading IMO. Would not something `makeExporterFromEnv` be better?

MikeGoldsmith commented:
@MrAlias @dashpole I've updated to use a SpanExporter factory instead of maintaining SpanExporter instances. Let me know what you think.

pdelewski commented:
LGTM

JamieDanielson commented on exporters/autoexport/exporter_test.go:
```diff
@@ -20,41 +20,37 @@ import (
 )
 
-func TestOTLPExporterIsReturnedWhenNoEnvOrFallbackExporterIsConfigured(t *testing.T) {
+func Test_otlp_exporter_returned_when_no_env_or_fallback_exported_configured(t *testing.T) {
```
nit: is there a reason the test names were changed from pascal to snake case? From what I can tell the former is more common throughout this repo.

MikeGoldsmith replied:
I found them much easier to read with such long names. I can revert if necessary but with them being test names, I thought clarity was important.

MikeGoldsmith replied:
Reverted test names to PascalCase.

JamieDanielson commented on exporters/autoexport/registry.go:
```diff
@@ -0,0 +1,151 @@
+
+	// errUnknownExpoter is returned when an unknown exporter name is used in
+	// the OTEL_*_EXPORTER environment variables.
+	errUnknownExpoter = errors.New("unknown exporter")
```
```suggestion
	errUnknownExporter = errors.New("unknown exporter")
```

JamieDanielson commented on exporters/autoexport/registry.go:
```diff
@@ -0,0 +1,151 @@
+	// and their mapping to a SpanExporter factory func() (trace.SpanExporter, error).
+	envRegistry = newRegistry()
+
+	// errUnknownExpoter is returned when an unknown exporter name is used in
```
```suggestion
	// errUnknownExporter is returned when an unknown exporter name is used in
```

JamieDanielson commented on exporters/autoexport/registry.go:
```diff
@@ -0,0 +1,151 @@
+
+// load returns tries to find the SpanExporter factory wioth the key and
+// then execute the factory, returning the created SpanExporter.
+// errUnknownExpoter is returned if the registration is missing and the error from
```
```suggestion
// errUnknownExporter is returned if the registration is missing and the error from
```

JamieDanielson commented on exporters/autoexport/registry.go:
```diff
@@ -0,0 +1,151 @@
+	defer r.mu.Unlock()
+	factory, ok := r.names[key]
+	if !ok {
+		return nil, errUnknownExpoter
```
```suggestion
		return nil, errUnknownExporter
```

JamieDanielson commented on exporters/autoexport/exporter.go:
```diff
@@ -0,0 +1,84 @@
+type Option func(*autoExportConfig)
+
+// WithFallabckSpanExporter sets the fallback exporter to use when no exporter
+// is configured through the OTEL_TRACES_EXPORTER environment vaiable.
```
```suggestion
// is configured through the OTEL_TRACES_EXPORTER environment variable.
```

JamieDanielson commented on exporters/autoexport/registry_test.go:
```diff
@@ -0,0 +1,117 @@
+	r := newRegistry()
+	assert.NotPanics(t, func() {
+		exp, err := r.load("non-existent")
+		assert.Equal(t, err, errUnknownExpoter, "empty registry should hold nothing")
```
```suggestion
		assert.Equal(t, err, errUnknownExporter, "empty registry should hold nothing")
```

JamieDanielson commented on exporters/autoexport/exporter.go:
```diff
@@ -0,0 +1,84 @@
+// Option applies an autoexport configuration option.
+type Option func(*autoExportConfig)
+
+// WithFallabckSpanExporter sets the fallback exporter to use when no exporter
```
```suggestion
// WithFallbackSpanExporter sets the fallback exporter to use when no exporter
```

JamieDanielson commented on exporters/autoexport/exporter.go:
```diff
@@ -0,0 +1,84 @@
+
+// NewTraceExporter returns a configured SpanExporter defined using the environment
+// variable OTEL_TRACES_EXPORTER, the configured fallback exporter via options or
+// a default OTLP expoter (in this order).
```
```suggestion
// a default OTLP exporter (in this order).
```

JamieDanielson commented on exporters/autoexport/exporter_test.go:
```diff
@@ -0,0 +1,59 @@
+	assert.Equal(t, testExporter, exporter)
+}
+
+func TestEnvExporterIsPrefereredOverFallbackExporter(t *testing.T) {
```
```suggestion
func TestEnvExporterIsPreferredOverFallbackExporter(t *testing.T) {
```

JamieDanielson commented on exporters/autoexport/exporter_test.go:
```diff
@@ -0,0 +1,59 @@
+	"go.opentelemetry.io/otel/sdk/trace"
+)
+
+func TestOTLPExporterReturnedWhenNoEnvOrFallbackExportedConfigured(t *testing.T) {
```
```suggestion
func TestOTLPExporterReturnedWhenNoEnvOrFallbackExporterConfigured(t *testing.T) {
```

JamieDanielson reviewed (commented):
🙈 Suggested a few typo updates that should cover all the usages too - sorry for the drive-by!  🚗

MrAlias commented on exporters/autoexport/registry.go:
```diff
@@ -0,0 +1,151 @@
+// RegisterSpanExporter sets the SpanExporter factory to be used when the
+// OTEL_TRACES_EXPORTERS environment variable contains the exporter name. This
+// will panic if name has already been registered.
+func RegisterSpanExporter(name string, factory func() (trace.SpanExporter, error)) {
```
The factory function should take a context. This will allow user values and timeouts to be passed to the factory in the future.

MrAlias commented on exporters/autoexport/registry.go:
```diff
@@ -0,0 +1,151 @@
+// already registered SpanExporter. A default OTLP exporter is registered
+// under both an empty string "" and "otlp".
+// An error is returned for any unknown exporters.
+func SpanExporter(name string) (trace.SpanExporter, error) {
```
Similar here, accepting a context will allow user customization that can be passed to the exporter.

MrAlias commented on exporters/autoexport/registry.go:
```diff
@@ -0,0 +1,151 @@
+
+	switch proto {
+	case "grpc":
+		return otlptracegrpc.New(context.Background())
```
These context would then come from the "top" level.

MrAlias reviewed (commented):
Mostly API tweaks.

The test look great 🏅

pellared commented on exporters/autoexport/exporter.go:
```diff
@@ -0,0 +1,84 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package autoexport // import "go.opentelemetry.io/contrib/exporters/autoexport"
```
Package comment (usually living in `doc.go`) is missing.

PS. I think this should be caught by `golangci-lint`.

pellared commented on exporters/autoexport/exporter.go:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	// if no env or fallback exporter, use OTLP exporter
+	exp, err := SpanExporter("otlp")
+	if err != nil {
+		otel.Handle(err)
+	}
```
I think it may be better to return `(trace.SpanExporter, error)`. Reasons:
1. If there is an error I think it would be rather important for the caller to handle it. Maybe the caller may want even to crash the application to make sure that telemetry pipeline is always working,\
2. This would be inline with the design of `autoprop`

pellared reviewed (commented):
Nice PR 👍 

You can also consider (not required) creating a `README.md`. See https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md#documentation for guidance.

MikeGoldsmith commented:
@MrAlias @pellared I've addressed your feedback. Please re-review 👍🏻

pellared commented on exporters/autoexport/README.md:
````diff
@@ -0,0 +1,56 @@
+exp, err := autoexport.NewTraceExporter(context.Background())
+```
+
+### Defining a custom fallback exporter
+
+The fallback exporter is returned if the exporter environment variable is not set.
+
+```golang
+import "go.opentelemetry.io/contrib/exporters/autoexport"
+
+fallbackExp = ...
+exp, err := autoexport.NewTraceExporter(
+    context.Background(),
+    WithFallbackSpanExporter(fallbackExp),
+)
+```
+
````
I feel this is redundant. We already have 

> If the environment variable is not set, the fallback exporter is returned. The fallback exporter defaults to an [OTLP exporter](https://github.com/open-telemetry/opentelemetry-go/tree/main/exporters/otlp/otlptrace) and can be overriden using the `WithFallbackSpanExporter` option.

pellared commented on exporters/autoexport/README.md:
```diff
@@ -0,0 +1,56 @@
+
+This module provides easy access to configuring a trace exporter that can be used when configuring an OpenTelemetry Go SDK trace export pipeline.
+
+The autoexport package looks for the `OTEL_TRACES_EXPORTER` environment variable and if set, attempts to load the exporter from it's registry of exporters. The registry is always loaded with an OTLP exporter with the key `"otlp"` and additional exporters can be registered using `autoexport.RegisterSpanExporter("name", ...)`(See example below). Exporter registration uses a factory pattern to not u ...
```
The functions should have a hyperlink to https://pkg.go.dev/

pellared commented on exporters/autoexport/README.md:
```diff
@@ -0,0 +1,56 @@
+
+The autoexport package looks for the `OTEL_TRACES_EXPORTER` environment variable and if set, attempts to load the exporter from it's registry of exporters. The registry is always loaded with an OTLP exporter with the key `"otlp"` and additional exporters can be registered using `autoexport.RegisterSpanExporter("name", ...)`(See example below). Exporter registration uses a factory pattern to not u ...
+
+If the environment variable is not set, the fallback exporter is returned. The fallback exporter defaults to an [OTLP exporter](https://github.com/open-telemetry/opentelemetry-go/tree/main/exporters/otlp/otlptrace) and can be overriden using the `WithFallbackSpanExporter` option.
```
The hyperlinks should direct to https://pkg.go.dev/ (not GitHub)

pellared replied:
Moreover the `WithFallbackSpanExporter` function should have a hyperlink to https://pkg.go.dev/

pellared commented on exporters/autoexport/exporter.go:
```diff
@@ -0,0 +1,98 @@
+// NewTraceExporter returns a configured SpanExporter defined using the environment
+// variable OTEL_TRACES_EXPORTER, the configured fallback exporter via options or
+// a default OTLP exporter (in this order).
+func NewTraceExporter(ctx context.Context, opts ...Option) (trace.SpanExporter, error) {
```
I am concerned about naming of  `NewTraceExporter` and `SpanExporter`.

1. How we will name the counterparts for metrics to handle `OTEL_METRICS_EXPORTER` in future? `NewMetricsExporter` and `MetricsExporter`?
2. By reading the names it is hard to understand which one uses the env var and which only uses the registry.
3. `TraceExporter` is never used in OTel Go API. I suggest keeping only SpanExporter to avoid confusion

Maybe we should rename `NewTraceExporter` to `SpanExporterFromEnv` (or `NewSpanExporter`)?

PS. I think more opinions are needed than just mine. Naming is hard 😉

pellared replied:
Side note: This is currently, the only reason why I have not approved the PR.

MikeGoldsmith replied:
Yes, I agree. I think `NewTraceExporter` was left over from the initial design. I've renamed to `NewSpanExporter`.

I also un-exported `registry.SpanExporter` func, users shouldn't try to load exporter's directly and should use the above func insteads.

@MrAlias Do you agree with this?

MrAlias replied:
> I also un-exported `registry.SpanExporter` func, users shouldn't try to load exporter's directly and should use the above func insteads.

We saw the case with the propagators that some users, like the collector, have a file configuration which includes a string defining the propagator they want to load. Having `registry.SpanExporter` would help those users not have to re-parse the string into an exporter here.

That said, we can add it afterwards like we did for the autoprop package.

pellared commented on exporters/autoexport/doc.go:
```diff
@@ -0,0 +1,18 @@
+
+// Package autoexport helps build trace exporters for configuring the
+// OpenTelemetry SDK trace export pipeline.
+
```
```suggestion
```

This will fix the lint error.

pellared commented on exporters/autoexport/README.md:
```diff
@@ -0,0 +1,56 @@
+# Automatic Exporter configuration
```
Side note: I just created https://github.com/open-telemetry/opentelemetry-go/pull/4136 😉

MikeGoldsmith commented:
@pellared How's this?

pellared commented on exporters/autoexport/doc.go:
```diff
@@ -12,6 +12,22 @@
+// variable and if set, attempts to load the exporter from it's registry of
+// exporters. The registry is always loaded with an OTLP exporter with the key
+// `"otlp"` and additional exporters can be registered using
+// [RegisterSpanExporter](https://pkg.go.dev/go.opentelemetry.io/contrib/exporters/autoexport#RegisterSpanExporter).
```
```suggestion
// [RegisterSpanExporter].
```

pellared commented on exporters/autoexport/doc.go:
```diff
@@ -12,6 +12,22 @@
+// The autoexport package looks for the `OTEL_TRACES_EXPORTER` environment
+// variable and if set, attempts to load the exporter from it's registry of
+// exporters. The registry is always loaded with an OTLP exporter with the key
+// `"otlp"` and additional exporters can be registered using
```
```suggestion
// `otlp` and additional exporters can be registered using
```

pellared commented on exporters/autoexport/doc.go:
```diff
@@ -12,6 +12,22 @@
+// If the environment variable is not set, the fallback exporter is returned.
+// The fallback exporter defaults to an
+// [OTLP exporter](https://pkg.go.dev/go.opentelemetry.io/otel/exporters/otlp/otlptrace)
+// and can be overriden using the [RegisterSpanExporter](https://pkg.go.dev/go.opentelemetry.io/contrib/exporters/autoexport#WithFallbackSpanExporter)
```
```suggestion
// and can be overridden using the [WithFallbackSpanExporter]
```

pellared replied:
Code review race condition 😄 
Addressed https://github.com/open-telemetry/opentelemetry-go-contrib/pull/2753/commits/f5fd3166f7ca07004e88b3daa030aed501c0412a

pellared commented on exporters/autoexport/doc.go:
```diff
@@ -12,6 +12,22 @@
+// that can be used when configuring an OpenTelemetry Go SDK trace export
+// pipeline.
+//
+// The autoexport package looks for the `OTEL_TRACES_EXPORTER` environment
+// variable and if set, attempts to load the exporter from it's registry of
```
```suggestion
// [NewTraceExporter] looks for the `OTEL_TRACES_EXPORTER` environment
// variable and if set, attempts to load the exporter from it's registry of
```

MrAlias commented on exporters/autoexport/registry.go:
```diff
@@ -0,0 +1,149 @@
+
+// store sets the factory for a key if is not already in the registry. errDuplicateRegistration
+// is returned if the registry already contains key.
+func (r *registry) store(key string, factory traceSpanExporterFactoryFunc) error {
```
`traceSpanExporterFactoryFunc` should be exported if we expect users to provide their own implementations of it as arguments.

pellared replied:
How about renaming to `SpanExporterFactory`?

MikeGoldsmith replied:
Removed the custom func type -- it's not necessary and was only used in a few places.

james-callahan commented:
Is there a no-op exporter we can use with autoexport?

pellared commented:
> Is there a no-op exporter we can use with autoexport?

I created https://github.com/open-telemetry/opentelemetry-go-contrib/issues/4013
